[TensorFlow Java] Fix iterable bug in generated code

samdow opened the pull request:
The code generated in TensorFlow Java (for instance Merge and Concat) is incorrect because the generated files expect a single input where the C++ expects an Input List. This is because the iterable boolean in TypeOf gets reset during the call to TypesOf. By moving the original if statement lower, we fix this bug without changing the intent of the code.

Commit:
Fix iterable bug

Commit:
Fix iterable bug

Co-authored-by: melissagrueter <melissagrueter@fb.com>
Co-authored-by: irenedea <irenedea@fb.com>

irenedea commented:
Issue here: https://github.com/tensorflow/tensorflow/issues/24483

Commit:
Merge branch 'master' of github.com:samdow/tensorflow

asimshankar reviewed (changes requested):
The fix looks good, thanks very much for it.
Could you add a unittest for this? You can add one to the newly added [`GeneratedOperationsTest.java`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/java/src/test/java/org/tensorflow/op/core/GeneratedOperationsTest.java), perhaps something like:

```java
@Test
public void testListInputTensorOutput() {
  // Something that uses ops.math().addN()
}
```

Thanks!

Commit:
Merge remote-tracking branch 'upstream/master'

Commit:
Added unit test for InputList fix

@Co-authored-by: irenedea<irenedea@fb.com>
@Co-authored-by: melissagrueter<melissagrueter@fb.com>

samdow commented:
Thanks @asimshankar for the quick reply and for setting up the framework for testing this! Let me know how it looks now

asimshankar commented on tensorflow/java/src/test/java/org/tensorflow/op/core/GeneratedOperationsTest.java:
```diff
@@ -40,4 +41,20 @@ public void tensorInputTensorOutput() {
+      inputs.add(ops.constant(1));
+      inputs.add(ops.constant(2));
+      inputs.add(ops.constant(3));
+      Operand<Integer> x = ops.math().addN(inputs);
```
Just to make sure I understand, we didn't _need_ `ArrayList`, right? This could have been:

```java
Ops ops = Ops.create(g);
Operand<Integer> x = ops.math().addN(new Operand<Integer>{ops.constant(1), ops.constant(2), ops.constant(3)});
```

?

samdow replied:
We were talking about this a bit and at least when I tried that didn't work (assuming you mean for it to be an array of Operands). This is because the inputs that are marked as iterable in the call to TypeOf are generated to take an [Iterable](https://docs.oracle.com/javase/8/docs/api/java/lang/Iterable.html), so I get an error
```
required: Iterable<Operand<T>>
found: Operand<Integer>[]
```
My Java is definitely a little out of date and I might have misunderstood the question, so I could definitely be missing something. We can also update the generate code more if this isn't the intended interface.

CC @melissagrueter @irenedea

asimshankar replied:
Ah, righto. Okay, fine for now (this PR should be merged shortly, we can revisit this later if needed).

asimshankar reviewed (approved):
Thanks!
